exec/java-exec/src/main/java/org/apache/drill/exec/store/dfs/easy/EasyGroupScan.java

```diff
@@ -333,21 +338,73 @@ public boolean canPushdownProjects(List<SchemaPath> columns) {
   }
 
   @JsonProperty
+  @JsonIgnore(value = false)
   public TupleMetadata getSchema() {
     return getTableMetadata().getSchema();
   }
 
   @Override
-  @JsonIgnore
-  public TableMetadata getTableMetadata() {
-    if (tableMetadata == null) {
-      tableMetadata = metadataProvider.getTableMetadata();
+  public AnalyzeInfoProvider getAnalyzeInfoProvider() {
+    return new AnalyzeFileInfoProviderImpl(formatPlugin.getName());
+  }
+
+  @Override
+  protected GroupScanWithMetadataFilterer<?> getFilterer() {
+    return new EasyGroupScanFilterer(this);
+  }
+
+  @Override
+  protected FileTableMetadataProviderBuilder<?> tableMetadataProviderBuilder(MetadataProviderManager source) {
+    if (source.usesMetastore()) {
+      return new MetastoreFileTableMetadataProvider.Builder<>((MetastoreMetadataProviderManager) source);
+    } else {
+      return defaultTableMetadataProviderBuilder(source);
     }
-    return tableMetadata;
   }
 
   @Override
-  public TableMetadataProvider getMetadataProvider() {
-    return metadataProvider;
+  protected FileTableMetadataProviderBuilder<?> defaultTableMetadataProviderBuilder(MetadataProviderManager source) {
+    return new SimpleFileTableMetadataProvider.Builder(source);
+  }
+
+  /**
+   * Implementation of GroupScanWithMetadataFilterer which uses {@link EasyGroupScan} as source and
+   * builds {@link EasyGroupScan} instance with filtered metadata.
+   */
+  private static class EasyGroupScanFilterer extends GroupScanWithMetadataFilterer<EasyGroupScanFilterer> {
+
+    EasyGroupScanFilterer(EasyGroupScan source) {
+      super(source);
+    }
+
+    @Override
+    public AbstractGroupScanWithMetadata<?> build() {
+      EasyGroupScan newScan = new EasyGroupScan((EasyGroupScan) source);
+      newScan.tableMetadata = tableMetadata;
+      // updates common row count and nulls counts for every column
+      if (newScan.getTableMetadata() != null && files != null && newScan.getFilesMetadata().size() != files.size()) {
+        newScan.tableMetadata = TableMetadataUtils.updateRowCount(newScan.getTableMetadata(), files.values());
+      }
+      newScan.partitions = partitions;
+      newScan.segments = segments;
+      newScan.files = files;
+      newScan.matchAllMetadata = matchAllMetadata;
+      newScan.nonInterestingColumnsMetadata = nonInterestingColumnsMetadata;
```

Comment:
Can you wrap the metadata items in a single object so you can pass it into the constructor or a new group scan instance, so the group scan itself is (mostly) immutable?

Reply:
Agreed. Elsewhere, we copy the group scan by passing the new value (filters, columns, etc.) into a constructor. Passing a half-dozen fields would be awkward. But, passing a single `MetadataSpec` (or whatever) would be clean:

```
private MyGroupScan(MyGroupScan from, MetadataSpec mdSpec) {
  this.foo = from.foo;
  ...
  this.metadata = mdSpec;
}

public MyGroupScan newWithMetadata(MetadataSpec mdSpec) {
  return new MyGroupScan(this, mdSpec);
}
```